server/resp/src/main/java/org/infinispan/server/resp/commands/connection/HELLO.java

```diff
@@ -35,25 +49,33 @@ public CompletionStage<RespRequestHandler> perform(Resp3AuthHandler handler,
       String version = new String(respProtocolBytes, CharsetUtil.UTF_8);
       if (!version.equals("3")) {
          ByteBufferUtils.stringToByteBufAscii("-NOPROTO sorry this protocol version is not supported\r\n", handler.allocator());
+         return handler.myStage();
+      }
+
+      if (arguments.size() == 4) {
+         successStage = handler.performAuth(ctx, arguments.get(2), arguments.get(3));
+      } else if (!handler.isAuthorized() && handler.canUseCertAuth()) {
+         successStage = handler.performAuth(ctx);
       } else {
-         if (arguments.size() == 4) {
-            successStage = handler.performAuth(ctx, arguments.get(2), arguments.get(3));
-         } else if (!handler.isAuthorized() && handler.canUseCertAuth()) {
-            successStage = handler.performAuth(ctx);
+         if (handler.isAuthenticationRequired()) {
+            RespErrorUtil.customRawError("NOAUTH HELLO must be called with the client already authenticated, otherwise the HELLO <proto> AUTH <user> <pass> option can be used to authenticate the client and select the RESP protocol version at the same time", handler.allocator());
          } else {
             helloResponse(handler.allocator());
          }
       }
 
       if (successStage != null) {
-         return handler.stageToReturn(successStage, ctx, success -> AUTH.createAfterAuthentication(success, handler));
+         return handler.stageToReturn(successStage.thenApply(s -> {
+            if (s) helloResponse(handler.allocator());
+            return s;
+         }), ctx, success -> AUTH.createAfterAuthentication(success, handler));
       }
 
       return handler.myStage();
    }
 
    private static void helloResponse(ByteBufPool alloc) {
-      String versionString = Version.getBrandVersion();
+      String versionString = Version.getMajorMinor();
```

Comment:
Man I really wish these were documented better as to what format you would expect.

Reply:
This is simply to work with Lettuce. Version 6.3.0 started parsing the version string returned and expects it to have only numbers: https://github.com/lettuce-io/lettuce-core/blob/main/src/main/java/io/lettuce/core/RedisHandshake.java#L322-L343

Let me add a comment there specifying that the version should be only numbers.

Reply:
I understand it is for Lettuce, I was more commenting wishing that our Version methods were documented with an example of what the String could look like.

We definitely need a comment here saying we need numbers only though.